src/ocean/net/ssl/SslClientConnection.d

```diff
@@ -0,0 +1,622 @@
+
+            if ( err ==  SSL_ERROR_WANT_WRITE )
+            {
+                this.select_client.ioWait(Event.EPOLLOUT);
+            }
+            else if (err == SSL_ERROR_WANT_READ)
+            {
+                this.select_client.ioWait(Event.EPOLLIN);
+            }
+            else
+            {
+                // The write failed
+
+                this.ssl_error("", "write", ERR_reason_error_string(err));
+            }
+        }
+
+        if ( w != request.length )
+        {
+            this.error("", "write", "Mismatch in number of bytes written");
+
+        }
+    }
+
+
+    /***************************************************************************
+
+        Reads a string from the SSL connection
+
+        The calling task is blocked until the read has completed.
+
+        Params:
+            buffer = array to store the string
+
+        Returns:
+            a slice into buffer of the bytes which were read.
+
+        Throws:
+            IOException if the read failed
+
+    ***************************************************************************/
+
+    public mstring read ( mstring buffer )
+    {
+        this.select_client.ioWait(Event.EPOLLIN);
+
+        ERR_clear_error();
+
+        int r;
+
+        while ( (r = SSL_read(this.sslHandle, buffer.ptr,
+            cast(int)buffer.length)) <= 0 )
+        {
+            auto err = SSL_get_error(this.sslHandle, r);
+
+            // At any time, a renegotiation is possible, so a call to SSL_read
+            // can also cause write operations.
+
+            if ( err ==  SSL_ERROR_WANT_WRITE )
+            {
+                this.select_client.ioWait(Event.EPOLLOUT);
+            }
+            else if (err == SSL_ERROR_WANT_READ)
+            {
+                this.select_client.ioWait(Event.EPOLLIN);
+            }
+            else
+            {
+                // The read failed
+
+                this.ssl_error("", "read", ERR_reason_error_string(err));
+            }
+
+        }
+        return buffer[0 .. r];
+    }
+
+
+    /***************************************************************************
+
+        Validate the X509 certificate.
+
+        The relevant documents are RFC 5280 and RFC 6125.
+
+        Params:
+            host_name = the name of the host
+
+        Throws:
+            IOException if validation fails
+
+    ***************************************************************************/
+
+    public void validateCertificate (cstring host_name)
+    {
+        // Step 1. Verify that a server certificate was presented during
+        // negotiation.
+
+        X509* cert = SSL_get_peer_certificate(this.sslHandle);
+
+        if ( cert )
+        {
+            X509_free(cert); // Free the certificate immediately
+        }
+
+        if (!cert)
+        {
+            this.error(host_name, "SSL_get_peer_certificate",
+                "No certificate was presented");
+        }
```

Comment:
Suggest
```d
if (X509* cert = SSL_get_peer_certificate(this.sslHandle))
{
    X509_free(cert); // Free the certificate immediately
}
else
{
    this.error(host_name, "SSL_get_peer_certificate",
        "No certificate was presented");
}
```
This protects from accidentally using `cert` if it is `null` or after it has been deallocated.